src/datasourcewidgets.js

```diff
@@ -0,0 +1,296 @@
+        widget.setValue = local_setValue;
+
+        return widget;
+    }
+
+    // -------------------------------------------------------------------------
+    // The following are functions related to "fixture" type data sources only
+
+    /**
+     * @param options - Optionally pass in:
+     *      onOptionsLoaded - callback called when options are loaded
+     *      hasAdvancedEditor - enable advanced editor if true
+     *      getSource(mug) - Initializes the source for the advanced editor
+     *      setSource(source, mug) - Saves the source from the advanced editor
+     */
+    function fixtureWidget(mug, options, labelText) {
+        var CUSTOM_XML = "Lookup table was not found in the project",
+            EMPTY_VALUE = JSON.stringify({src: "", id: "", query: ""});
+
+        function isEmptyValue(val) {
+            return !val || _.all(_.map(val, _.isEmpty));
+        }
+
+        function local_getValue() {
+            return JSON.parse(super_getValue());
+        }
+
+        function local_setValue(val) {
+            var jsonVal = EMPTY_VALUE;
+            if (!isEmptyValue(val)) {
+                jsonVal = JSON.stringify(val);
+                if (!widget.equivalentOption(jsonVal)) {
+                    widget.addOption(jsonVal, CUSTOM_XML);
+                }
+            }
+            super_setValue(jsonVal);
+            hasValue = true;
+        }
+
+        var widget = widgets.dropdown(mug, options), 
+            super_getValue = widget.getValue,
+            super_setValue = widget.setValue,
+            getSource = options.getSource ? options.getSource : local_getValue,
+            setSource = options.setSource ? options.setSource : local_setValue,
+            hasValue = false;
+
+        widget.addOption(EMPTY_VALUE, "Loading...");
+        datasources.getDataSources(function (data) {
+            var value;
+            if (options.dataSourcesFilter) {
+                data = options.dataSourcesFilter(data);
+            }
+            if (hasValue) {
+                value = local_getValue();
+            }
+            widget.clearOptions();
+            var fixtures = getPossibleFixtures(data);
+            widget.addOptions(generateFixtureOptions(fixtures));
+            if (hasValue && !isEmptyValue(value)) {
+                local_setValue(value);
+            } else if (fixtures && fixtures.length) {
+                // default to first option
+                local_setValue(_.omit(fixtures[0], 'name'));
+            }
+            if (options.onOptionsLoaded) {
+                options.onOptionsLoaded(data);
+            }
+        });
+
+        if (options.hasAdvancedEditor) {
+            widget.getUIElement = function () {
+                var query = widgets.util.getUIElementWithEditButton(
+                        widgets.util.getUIElement(widget.input, labelText),
+                        function () {
+                            mug.form.vellum.displaySecondaryEditor({
```

Comment:
Had to change this from `vellum` to `mug.form.vellum` I don't think that there should ever an instance where `mug.form.vellum` doesn't exist (and tests seem to agree with me)